File: Tests/ConnectorControlStateTests/ImportModelTests.swift
import XCTest
import ConnectorControlCore
import ConnectorControlTestSupport
@testable import ConnectorControlState

/// windows/tests/ConnectorControl.Core.Tests/State/ImportModelTests.cs. The Import sheet: what
/// the rows say about one document against the collection it would land in, what the count
/// follows, and what a document this app cannot read leaves on screen.
@MainActor
final class ImportModelTests: XCTestCase {
    private func write(_ doc: CollectionDocument, at url: URL) throws {
        try FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try doc.serialized().write(to: url)
    }

    func testRowsShowCollisionsAndTheCountFollowsChoices() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        let url = h.dir.file("shared/data-team.json")
        try write(CollectionDocumentSamples.dataTeam, at: url)
        // One of the document's four connectors is already in the target, under the same name.
        XCTAssertNil(state.upsert(name: "github", entry: MCPEntry(enabled: true, config: AppStateHarness.remote("https://x/")),
                                  renamedFrom: nil))

        let model = ImportModel(state: state, path: url.path)
        XCTAssertNil(model.loadError)
        XCTAssertEqual(model.mode, .addToCollection)
        XCTAssertEqual(model.targetCollection, "Default", "the active collection is local, so it is the target")
        XCTAssertEqual(model.localCollections, ["Default"])
        XCTAssertEqual(model.sourceLine, ImportModel.sourceLine("Data team", "Acme Data Platform", 4))
        XCTAssertEqual(model.rows.map(\.name), ["dbt", "github", "ledger", "notion"])
        XCTAssertEqual(model.rows.map(\.present), [false, true, false, false])
        XCTAssertEqual(model.rows.map(\.include), [true, false, true, true], "what is already there is not imported by default")
        XCTAssertEqual(model.rows.map(\.choice), [.add, .replace, .add, .add])
        XCTAssertEqual(model.rows.first { $0.name == "ledger" }?.needs, ["server_path"])
        XCTAssertTrue(model.rows.allSatisfy { $0.excludedReason == nil }, "a Mac excludes nothing")
        XCTAssertEqual(model.importCount, 3)
        XCTAssertTrue(model.canImport)

        // Ticking the collision in imports it too; the count follows the ticks, not the document.
        model.rows[1].include = true
        XCTAssertEqual(model.importCount, 4)
        model.rows[0].include = false
        XCTAssertEqual(model.importCount, 3)

        model.rows[1].choice = .keepBoth
        XCTAssertNil(model.perform())
        let mcps = try XCTUnwrap(state.store.collections["Default"]).mcps
        XCTAssertNil(mcps["dbt"], "an unticked row is skipped")
        XCTAssertNotNil(mcps["github 2"], "Keep both lands beside the connector that was there")
        XCTAssertEqual(mcps["github"]?.config, AppStateHarness.remote("https://x/"), "the one that was there is untouched")
        XCTAssertEqual(state.collectionsFile.collections["Default"]?.provenance["ledger"]?.from, "Data team")
        XCTAssertEqual(state.kind(of: "Default"), .local)
    }

    /// The platform-forced half of a mirrored pair: the Mac never writes the `cmd /c` launcher,
    /// so a header name cmd.exe would re-parse excludes nothing here, where the Windows mirror
    /// asserts the row is excluded, uncountable and skipped.
    func testAnExcludedRowShowsItsReasonStaysOutOfTheCountAndIsSkipped() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        let url = h.dir.file("shared/risky.json")
        try write(riskyHeaderDocument, at: url)

        let model = ImportModel(state: state, path: url.path)
        XCTAssertEqual(model.rows.map(\.name), ["bad", "good"])
        XCTAssertTrue(model.rows.allSatisfy { $0.excludedReason == nil })
        XCTAssertTrue(model.rows.allSatisfy(\.include))
        XCTAssertEqual(model.importCount, 2)
        model.mode = .keepInSync
        XCTAssertEqual(model.importCount, 2)

        model.mode = .addToCollection
        XCTAssertNil(model.perform())
        let mcps = try XCTUnwrap(state.store.collections["Default"]).mcps
        XCTAssertNotNil(mcps["good"])
        XCTAssertNotNil(mcps["bad"], "a header name is only unsafe where cmd.exe re-parses it")
    }

    /// Two remote connectors, one with a header name carrying the `&` the Windows `cmd /c`
    /// launcher cannot hand to cmd.exe.
    private var riskyHeaderDocument: CollectionDocument {
        CollectionDocument(
            name: "Risky", author: "Acme", origin: "o-risky", exported: "2026-09-21T14:02:11Z",
            connectors: [
                "bad": .init(launcher: .remote(.init(url: "https://h/mcp", auth: .header(name: "X&Y"),
                                                     package: "mcp-remote", extraArgs: []))),
                "good": .init(launcher: .remote(.init(url: "https://h/mcp", auth: .automatic,
                                                      package: "mcp-remote", extraArgs: []))),
            ])
    }

    func testSyncModeDefaultsTheNameAndSuffixesATakenOne() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        let url = h.dir.file("shared/data-team.json")
        try write(CollectionDocumentSamples.dataTeam, at: url)

        let first = ImportModel(state: state, path: url.path)
        XCTAssertEqual(first.syncName, "Data team", "the document's own name is the default")
        first.mode = .keepInSync
        XCTAssertEqual(first.importCount, 4, "sync mode takes every connector this platform can carry")
        XCTAssertTrue(first.canImport)
        XCTAssertNil(first.perform())
        XCTAssertEqual(state.kind(of: "Data team"), .synced)
        XCTAssertEqual(state.sourceBinding(of: "Data team")?.path, url.standardizedFileURL.path)

        let second = ImportModel(state: state, path: url.path)
        XCTAssertEqual(second.syncName, "Data team 2", "a name already taken is suffixed")
        second.mode = .keepInSync
        second.syncName = "  "
        XCTAssertFalse(second.canImport, "a collection has to be called something")
    }

    func testANewerDocumentIsALoadError() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        var newer = CollectionDocumentSamples.dataTeam.encode()
        newer = try XCTUnwrap(newer.replacing(at: JSONPointer(["connectorControlCollection"]), with: .int(2)))
        let url = h.dir.file("future.json")
        try newer.serialized().write(to: url)

        let model = ImportModel(state: state, path: url.path)
        XCTAssertEqual(model.loadError, AppState.newerDocumentError)
        XCTAssertTrue(model.rows.isEmpty)
        XCTAssertEqual(model.importCount, 0)
        XCTAssertFalse(model.canImport)
        XCTAssertEqual(model.perform(), AppState.newerDocumentError, "the sheet's button says what the sheet says")

        let half = h.dir.file("half.json")
        try TempDir.touch(half, "{half")
        let malformed = ImportModel(state: state, path: half.path)
        XCTAssertEqual(malformed.loadError?.hasPrefix("half.json couldn’t be read: "), true)
        XCTAssertFalse(malformed.canImport)
        XCTAssertEqual(state.collectionNames, ["Default"], "a document that cannot be read creates nothing")
    }

    func testEveryChoiceHasATitleAndACollisionOffersThree() {
        XCTAssertEqual(ImportModel.choiceTitle(.add), "Add")
        XCTAssertEqual(ImportModel.choiceTitle(.replace), "Replace")
        XCTAssertEqual(ImportModel.choiceTitle(.keepBoth), "Keep both")
        XCTAssertEqual(ImportModel.choiceTitle(.skip), "Skip")
        // The picker is a collision's, so the case where nothing is in the way is not in it.
        XCTAssertEqual(ImportModel.collisionChoices, [.replace, .keepBoth, .skip])
        XCTAssertEqual(ImportModel.collisionChoices.map(ImportModel.choiceTitle),
                       ["Replace", "Keep both", "Skip"])
    }

    func testARowsCautionIsTheSentenceAConnectorAlreadyCarries() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        let url = h.dir.file("data-team.json")
        try write(CollectionDocumentSamples.dataTeam, at: url)
        let model = ImportModel(state: state, path: url.path)

        // The document's own needs, sorted, in the wording the window and the editor use.
        let dbt = try XCTUnwrap(model.rows.first { $0.name == "dbt" })
        XCTAssertEqual(dbt.needs, ["DBT_TOKEN"])
        XCTAssertEqual(dbt.needsCaution, AppState.needsValueCaution("DBT_TOKEN"))
        let notion = try XCTUnwrap(model.rows.first { $0.name == "notion" })
        XCTAssertEqual(notion.needs, ["token"])
        XCTAssertEqual(notion.needsCaution, AppState.needsValueCaution("token"))

        // No glyph for a connector the author left nothing to fill in.
        let github = try XCTUnwrap(model.rows.first { $0.name == "github" })
        XCTAssertEqual(github.needs, [])
        XCTAssertNil(github.needsCaution)

        // The same connector, once imported, says exactly the same thing in the window.
        XCTAssertNil(model.perform())
        XCTAssertEqual(state.connectorCaution("dbt", in: "Default"), dbt.needsCaution)
    }
    func testEachRowCarriesItsOwnBadge() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        let url = h.dir.file("data-team.json")
        try write(CollectionDocumentSamples.dataTeam, at: url)
        // One of the document's four connectors is already in the target, under the same name.
        XCTAssertNil(state.upsert(name: "github", entry: MCPEntry(config: AppStateHarness.remote("https://x/")),
                                  renamedFrom: nil))

        let model = ImportModel(state: state, path: url.path)
        XCTAssertEqual(model.rows.map(\.name), ["dbt", "github", "ledger", "notion"])
        XCTAssertEqual(model.rows.map(\.badge),
                       [ImportModel.newBadge, ImportModel.presentBadge, ImportModel.newBadge, ImportModel.newBadge])
        // The badge is what the row is, not what the user has since ticked.
        model.rows[1].include = true
        model.rows[1].choice = .replace
        XCTAssertEqual(model.rows[1].badge, ImportModel.presentBadge)
        // The third form is the excluded one, which only the other platform's launcher rules produce;
        // its mirror asserts it against a row this Mac cannot make.
    }
    /// Two needs in one connector, one in an argument and one in an environment value, so first
    /// appearance and alphabetical order disagree: object keys are walked sorted, and `args`
    /// comes before `env`, while the names sort the other way.
    private var twoNeedsDocument: CollectionDocument {
        CollectionDocument(
            name: "Two", author: "Acme", origin: "o-two", exported: "2026-09-21T14:02:11Z",
            connectors: [
                "pair": .init(launcher: .local(.init(command: "node", args: ["${CC_NEEDS:zulu}"], platform: .mac)),
                              env: ["ALPHA": .hint("the alpha hint")],
                              needs: ["zulu": "the zulu hint"]),
            ])
    }

    func testARowsNeedsFollowTheConfigRatherThanTheAlphabet() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        let url = h.dir.file("two.json")
        try write(twoNeedsDocument, at: url)
        let model = ImportModel(state: state, path: url.path)

        let pair = try XCTUnwrap(model.rows.first { $0.name == "pair" })
        XCTAssertEqual(pair.needs, ["zulu", "ALPHA"], "first appearance in the config, not sorted")
        XCTAssertEqual(pair.needsCaution, AppState.needsValueCaution("zulu, ALPHA"))

        // Once imported, the connector's own caution is the very same sentence.
        XCTAssertNil(model.perform())
        XCTAssertEqual(state.connectorCaution("pair", in: "Default"), pair.needsCaution)
    }
    func testATickedRowSetToSkipIsNotCounted() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        let url = h.dir.file("shared/data-team.json")
        try write(CollectionDocumentSamples.dataTeam, at: url)
        XCTAssertNil(state.upsert(name: "github", entry: MCPEntry(config: AppStateHarness.remote("https://x/")),
                                  renamedFrom: nil))

        let model = ImportModel(state: state, path: url.path)
        XCTAssertEqual(model.importCount, 3, "the collision starts unticked")

        // Ticked and set to Skip: the button must not promise what `perform` will not land.
        model.rows[1].include = true
        model.rows[1].choice = .skip
        XCTAssertEqual(model.importCount, 3)
        model.rows[1].choice = .replace
        XCTAssertEqual(model.importCount, 4)

        // Every row skipped is nothing to import at all.
        for index in model.rows.indices { model.rows[index].choice = .skip }
        XCTAssertEqual(model.importCount, 0)
        XCTAssertFalse(model.canImport)
        // Sync mode takes the whole document, so a per-row choice says nothing about it.
        model.mode = .keepInSync
        XCTAssertEqual(model.importCount, 4)
    }

    func testTheAccessibilityLabelsNameTheirControls() {
        XCTAssertEqual(ImportModel.includeLabel("github"), "Include github")
        XCTAssertEqual(ImportModel.syncNameLabel, "Collection name")
        XCTAssertEqual(ImportModel.collisionPickerLabel("github"), "What to do with github")
    }
}

File: windows/tests/ConnectorControl.Core.Tests/State/ImportModelTests.cs
using ConnectorControl.Core.State;
using ConnectorControl.Core.Tests.TestSupport;

namespace ConnectorControl.Core.Tests.State;

/// <summary>
/// Tests/ConnectorControlStateTests/ImportModelTests.swift. The Import sheet: what the rows say
/// about one document against the collection it would land in, what the count follows, and what a
/// document this app cannot read leaves on screen.
/// </summary>
public class ImportModelTests
{
    private static void Write(CollectionDocument doc, string path)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(path)!);
        File.WriteAllBytes(path, doc.Serialize());
    }

    [Fact]
    public void RowsShowCollisionsAndTheCountFollowsChoices()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var path = Path.Combine(h.Dir.File("shared"), "data-team.json");
        Write(CollectionDocumentSamples.DataTeam, path);
        // One of the document's four connectors is already in the target, under the same name.
        Assert.Null(state.Upsert("github", new McpEntry(true, AppStateHarness.Remote("https://x/")), null));

        var model = new ImportModel(state, path);
        Assert.Null(model.LoadError);
        Assert.Equal(ImportModel.Mode.AddToCollection, model.ImportMode);
        // The active collection is local, so it is the target.
        Assert.Equal("Default", model.TargetCollection);
        Assert.Equal(["Default"], model.LocalCollections);
        Assert.Equal(ImportModel.SourceLine("Data team", "Acme Data Platform", 4), model.SourceSentence);
        Assert.Equal(["dbt", "github", "ledger", "notion"], model.Rows.Select(r => r.Name));
        Assert.Equal([false, true, false, false], model.Rows.Select(r => r.Present));
        // What is already there is not imported by default.
        Assert.Equal([true, false, true, true], model.Rows.Select(r => r.Include));
        Assert.Equal([ImportChoice.Add, ImportChoice.Replace, ImportChoice.Add, ImportChoice.Add],
            model.Rows.Select(r => r.Choice));
        // An unticked collision shows its badge.
        Assert.Equal([false, false, false, false], model.Rows.Select(r => r.ShowsPicker));
        var raised = new List<string?>();
        model.Rows[1].PropertyChanged += (_, e) => raised.Add(e.PropertyName);
        model.Rows[1].Include = true;
        Assert.True(model.Rows[1].ShowsPicker);   // a collision coming across asks what to do
        Assert.Contains(nameof(ImportModel.Row.ShowsPicker), raised);
        model.Rows[1].Include = false;
        Assert.Equal(["server_path"], model.Rows.Single(r => r.Name == "ledger").Needs);
        // Nothing in the sample trips the cmd guard.
        Assert.All(model.Rows, row => Assert.Null(row.ExcludedReason));
        Assert.Equal(3, model.ImportCount);
        Assert.True(model.CanImport);

        // Ticking the collision in imports it too; the count follows the ticks, not the document.
        model.Rows[1].Include = true;
        Assert.Equal(4, model.ImportCount);
        model.Rows[0].Include = false;
        Assert.Equal(3, model.ImportCount);

        model.Rows[1].Choice = ImportChoice.KeepBoth;
        Assert.Null(model.Perform());
        var mcps = state.Store.Collections["Default"].Mcps;
        Assert.False(mcps.ContainsKey("dbt"), "an unticked row is skipped");
        Assert.True(mcps.ContainsKey("github 2"), "Keep both lands beside the connector that was there");
        // The one that was there is untouched.
        Assert.Equal(AppStateHarness.Remote("https://x/"), mcps["github"].Config);
        Assert.Equal("Data team", state.CollectionsFile.Collections["Default"].Provenance["ledger"].From);
        Assert.Equal(CollectionKind.Local, state.KindOf("Default"));
    }

    /// <summary>
    /// Two remote connectors, one with a header name carrying the &amp; the Windows cmd /c
    /// launcher cannot hand to cmd.exe.
    /// </summary>
    private static CollectionDocument RiskyHeaderDocument() => new(
        "Risky", "Acme", "o-risky", "2026-09-21T14:02:11Z",
        new Dictionary<string, CollectionDocument.Connector>
        {
            ["bad"] = new(new CollectionDocument.Launcher.Remote(
                "https://h/mcp", new CollectionDocument.Auth.Header("X&Y"), "mcp-remote", [])),
            ["good"] = new(new CollectionDocument.Launcher.Remote(
                "https://h/mcp", CollectionDocument.Auth.Auto, "mcp-remote", [])),
        });

    /// <summary>
    /// The platform-forced half of a mirrored pair: only this build writes the cmd /c launcher,
    /// so the Mac mirror asserts the same document excludes nothing and every row counts.
    /// </summary>
    [Fact]
    public void AnExcludedRowShowsItsReasonStaysOutOfTheCountAndIsSkipped()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var path = Path.Combine(h.Dir.File("shared"), "risky.json");
        Write(RiskyHeaderDocument(), path);

        var model = new ImportModel(state, path);
        Assert.Equal(["bad", "good"], model.Rows.Select(r => r.Name));
        var bad = model.Rows[0];
        Assert.Equal(RemotePattern.CmdUnsafeReason(RemoteField.HeaderName), bad.ExcludedReason);
        Assert.Equal("skipped: " + RemotePattern.CmdUnsafeReason(RemoteField.HeaderName),
            ImportModel.SkippedBadge(bad.ExcludedReason!));
        // The row carries that text itself, so the template needs no converter over the factory.
        Assert.Equal(ImportModel.SkippedBadge(bad.ExcludedReason!), bad.Badge);
        Assert.False(bad.Include);
        Assert.Null(model.Rows[1].ExcludedReason);
        Assert.True(model.Rows[1].Include);

        // Out of the count in both modes: this machine has no way to run it either way.
        Assert.Equal(1, model.ImportCount);
        model.ImportMode = ImportModel.Mode.KeepInSync;
        Assert.Equal(1, model.ImportCount);
        model.ImportMode = ImportModel.Mode.AddToCollection;

        // Ticking it makes no difference — an excluded row cannot be included.
        bad.Include = true;
        Assert.Equal(1, model.ImportCount);
        Assert.Null(model.Perform());
        var mcps = state.Store.Collections["Default"].Mcps;
        Assert.True(mcps.ContainsKey("good"));
        Assert.False(mcps.ContainsKey("bad"));
        Assert.False(state.CollectionsFile.Collections["Default"].Provenance.ContainsKey("bad"));
    }

    [Fact]
    public void SyncModeDefaultsTheNameAndSuffixesATakenOne()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var path = Path.Combine(h.Dir.File("shared"), "data-team.json");
        Write(CollectionDocumentSamples.DataTeam, path);

        var first = new ImportModel(state, path);
        Assert.Equal("Data team", first.SyncName);   // the document's own name is the default
        first.ImportMode = ImportModel.Mode.KeepInSync;
        // Sync mode takes every connector this platform can carry.
        Assert.Equal(4, first.ImportCount);
        Assert.True(first.CanImport);
        Assert.Null(first.Perform());
        Assert.Equal(CollectionKind.Synced, state.KindOf("Data team"));
        Assert.Equal(path, state.SourceBinding("Data team")?.Path);

        var second = new ImportModel(state, path);
        Assert.Equal("Data team 2", second.SyncName);   // a name already taken is suffixed
        second.ImportMode = ImportModel.Mode.KeepInSync;
        second.SyncName = "  ";
        Assert.False(second.CanImport, "a collection has to be called something");
    }

    [Fact]
    public void ANewerDocumentIsALoadError()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var newer = CollectionDocumentSamples.DataTeam.Encode()
            .Replacing(JsonPointer.Parse("/connectorControlCollection")!, JsonValue.Int(2))!;
        var path = h.Dir.File("future.json");
        File.WriteAllBytes(path, newer.Serialize());

        var model = new ImportModel(state, path);
        Assert.Equal(AppState.NewerDocumentError, model.LoadError);
        Assert.Empty(model.Rows);
        Assert.Equal(0, model.ImportCount);
        Assert.False(model.CanImport);
        // The sheet's button says what the sheet says.
        Assert.Equal(AppState.NewerDocumentError, model.Perform());

        var half = h.Dir.File("half.json");
        File.WriteAllText(half, "{half");
        var malformed = new ImportModel(state, half);
        Assert.StartsWith("half.json couldn’t be read: ", malformed.LoadError, StringComparison.Ordinal);
        Assert.False(malformed.CanImport);
        // A document that cannot be read creates nothing.
        Assert.Equal(["Default"], state.CollectionNames);
    }

    [Fact]
    public void EveryChoiceHasATitleAndACollisionOffersThree()
    {
        Assert.Equal("Add", ImportModel.ChoiceTitle(ImportChoice.Add));
        Assert.Equal("Replace", ImportModel.ChoiceTitle(ImportChoice.Replace));
        Assert.Equal("Keep both", ImportModel.ChoiceTitle(ImportChoice.KeepBoth));
        Assert.Equal("Skip", ImportModel.ChoiceTitle(ImportChoice.Skip));
        // The picker is a collision's, so the case where nothing is in the way is not in it.
        Assert.Equal([ImportChoice.Replace, ImportChoice.KeepBoth, ImportChoice.Skip], ImportModel.CollisionChoices);
        Assert.Equal(["Replace", "Keep both", "Skip"], ImportModel.CollisionChoices.Select(ImportModel.ChoiceTitle));
    }

    [Fact]
    public void ARowsCautionIsTheSentenceAConnectorAlreadyCarries()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var path = h.Dir.File("data-team.json");
        Write(CollectionDocumentSamples.DataTeam, path);
        var model = new ImportModel(state, path);

        // The document's own needs, sorted, in the wording the window and the editor use.
        var dbt = model.Rows.Single(r => r.Name == "dbt");
        Assert.Equal(["DBT_TOKEN"], dbt.Needs);
        Assert.Equal(AppState.NeedsValueCaution("DBT_TOKEN"), dbt.NeedsCaution);
        var notion = model.Rows.Single(r => r.Name == "notion");
        Assert.Equal(["token"], notion.Needs);
        Assert.Equal(AppState.NeedsValueCaution("token"), notion.NeedsCaution);

        // No glyph for a connector the author left nothing to fill in.
        var github = model.Rows.Single(r => r.Name == "github");
        Assert.Empty(github.Needs);
        Assert.Null(github.NeedsCaution);

        // The same connector, once imported, says exactly the same thing in the window.
        Assert.Null(model.Perform());
        Assert.Equal(dbt.NeedsCaution, state.ConnectorCaution("dbt", "Default"));
    }
    [Fact]
    public void EachRowCarriesItsOwnBadge()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var path = h.Dir.File("data-team.json");
        Write(CollectionDocumentSamples.DataTeam, path);
        // One of the document's four connectors is already in the target, under the same name.
        Assert.Null(state.Upsert("github", new McpEntry(AppStateHarness.Remote("https://x/")), null));

        var model = new ImportModel(state, path);
        Assert.Equal(["dbt", "github", "ledger", "notion"], model.Rows.Select(r => r.Name));
        Assert.Equal(
            [ImportModel.NewBadge, ImportModel.PresentBadge, ImportModel.NewBadge, ImportModel.NewBadge],
            model.Rows.Select(r => r.Badge));
        // The badge is what the row is, not what the user has since ticked.
        model.Rows[1].Include = true;
        model.Rows[1].Choice = ImportChoice.Replace;
        Assert.Equal(ImportModel.PresentBadge, model.Rows[1].Badge);
    }
    /// <summary>
    /// Two needs in one connector, one in an argument and one in an environment value, so first
    /// appearance and alphabetical order disagree: object keys are walked sorted, and Args comes
    /// before Env, while the names sort the other way.
    /// </summary>
    private static CollectionDocument TwoNeedsDocument() => new(
        "Two", "Acme", "o-two", "2026-09-21T14:02:11Z",
        new Dictionary<string, CollectionDocument.Connector>
        {
            ["pair"] = new(
                new CollectionDocument.Launcher.Local("node", ["${CC_NEEDS:zulu}"], CollectionPlatform.Mac),
                env: [new("ALPHA", new CollectionDocument.EnvValue.Hint("the alpha hint"))],
                needs: [new("zulu", "the zulu hint")]),
        });

    [Fact]
    public void ARowsNeedsFollowTheConfigRatherThanTheAlphabet()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var path = h.Dir.File("two.json");
        Write(TwoNeedsDocument(), path);
        var model = new ImportModel(state, path);

        var pair = model.Rows.Single(r => r.Name == "pair");
        // First appearance in the config, not sorted.
        Assert.Equal(["zulu", "ALPHA"], pair.Needs);
        Assert.Equal(AppState.NeedsValueCaution("zulu, ALPHA"), pair.NeedsCaution);

        // Once imported, the connector's own caution is the very same sentence.
        Assert.Null(model.Perform());
        Assert.Equal(pair.NeedsCaution, state.ConnectorCaution("pair", "Default"));
    }
    [Fact]
    public void ATickedRowSetToSkipIsNotCounted()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var path = Path.Combine(h.Dir.File("shared"), "data-team.json");
        Write(CollectionDocumentSamples.DataTeam, path);
        Assert.Null(state.Upsert("github", new McpEntry(AppStateHarness.Remote("https://x/")), null));

        var model = new ImportModel(state, path);
        Assert.Equal(3, model.ImportCount);   // the collision starts unticked

        // Ticked and set to Skip: the button must not promise what Perform will not land.
        model.Rows[1].Include = true;
        model.Rows[1].Choice = ImportChoice.Skip;
        Assert.Equal(3, model.ImportCount);
        model.Rows[1].Choice = ImportChoice.Replace;
        Assert.Equal(4, model.ImportCount);

        // Every row skipped is nothing to import at all.
        foreach (var row in model.Rows)
        {
            row.Choice = ImportChoice.Skip;
        }
        Assert.Equal(0, model.ImportCount);
        Assert.False(model.CanImport);
        // Sync mode takes the whole document, so a per-row choice says nothing about it.
        model.ImportMode = ImportModel.Mode.KeepInSync;
        Assert.Equal(4, model.ImportCount);
    }

    /// <summary>
    /// C#-only: SwiftUI binds the optional <c>loadError</c> itself, so the Mac has no bool to
    /// assert — the same one-sided pattern <c>HasCollectionBanner</c> already follows.
    /// </summary>
    [Fact]
    public void ADocumentThatCannotBeReadSaysSoAsABool()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var good = Path.Combine(h.Dir.File("shared"), "data-team.json");
        Write(CollectionDocumentSamples.DataTeam, good);
        Assert.False(new ImportModel(state, good).HasLoadError);

        var half = h.Dir.File("half.json");
        File.WriteAllText(half, "{half");
        var malformed = new ImportModel(state, half);
        Assert.True(malformed.HasLoadError);
        Assert.NotNull(malformed.LoadError);
    }

    [Fact]
    public void TheAccessibilityLabelsNameTheirControls()
    {
        Assert.Equal("Include github", ImportModel.IncludeLabel("github"));
        Assert.Equal("Collection name", ImportModel.SyncNameLabel);
        Assert.Equal("What to do with github", ImportModel.CollisionPickerLabel("github"));
    }
    /// <summary>
    /// C#-only: the Mac's rows are structs in a @Published array, so editing one republishes the
    /// array and nothing model-side needs to raise.
    /// </summary>
    [Fact]
    public void TickingOrSkippingARowRaisesTheCountWithoutAViewCall()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var path = Path.Combine(h.Dir.File("shared"), "data-team.json");
        Write(CollectionDocumentSamples.DataTeam, path);
        Assert.Null(state.Upsert("github", new McpEntry(AppStateHarness.Remote("https://x/")), null));

        var model = new ImportModel(state, path);
        var raised = new List<string>();
        model.PropertyChanged += (_, e) => raised.Add(e.PropertyName ?? "");
        Assert.Equal(3, model.ImportCount);

        // A choice the sheet makes on a row has to reach the footer, which reads the model.
        model.Rows[1].Include = true;
        Assert.Contains(nameof(ImportModel.ImportCount), raised);
        Assert.Contains(nameof(ImportModel.CanImport), raised);
        Assert.Equal(4, model.ImportCount);

        raised.Clear();
        model.Rows[1].Choice = ImportChoice.Skip;
        Assert.Contains(nameof(ImportModel.ImportCount), raised);
        Assert.Equal(3, model.ImportCount);

        // Setting a row to what it already holds says nothing.
        raised.Clear();
        model.Rows[1].Choice = ImportChoice.Skip;
        Assert.Empty(raised);

        // Rebuilding the rows lets the old ones go: the replaced row no longer reaches the model.
        var stale = model.Rows[1];
        Assert.Null(state.CreateCollection("Other"));
        model.TargetCollection = "Other";
        raised.Clear();
        stale.Include = !stale.Include;
        Assert.Empty(raised);
    }

    /// <summary>
    /// C#-only, as the test above is: on the Mac the mode, the target and the name are @Published,
    /// so changing one republishes the model. Here the Import button's gate and count follow each
    /// of them, and the dialog must not have to nudge its own binding to see that.
    /// </summary>
    [Fact]
    public void TheModeTheTargetAndTheNameRaiseTheGateThatFollowsThem()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var path = Path.Combine(h.Dir.File("shared"), "data-team.json");
        Write(CollectionDocumentSamples.DataTeam, path);
        var model = new ImportModel(state, path);
        var raised = new List<string>();
        model.PropertyChanged += (_, e) => raised.Add(e.PropertyName ?? "");

        model.ImportMode = ImportModel.Mode.KeepInSync;
        Assert.Contains(nameof(ImportModel.CanImport), raised);
        Assert.Contains(nameof(ImportModel.ImportCount), raised);

        raised.Clear();
        model.SyncName = "  ";
        Assert.Contains(nameof(ImportModel.CanImport), raised);
        Assert.False(model.CanImport);

        raised.Clear();
        model.ImportMode = ImportModel.Mode.AddToCollection;
        Assert.Null(state.CreateCollection("Other"));
        raised.Clear();
        model.TargetCollection = "Other";
        Assert.Contains(nameof(ImportModel.CanImport), raised);
        Assert.Contains(nameof(ImportModel.ImportCount), raised);

        // Setting what is already there says nothing.
        raised.Clear();
        model.ImportMode = ImportModel.Mode.AddToCollection;
        model.SyncName = "  ";
        Assert.Empty(raised);
    }
}
